README.md
# pca3d
Three Dimensional PCA Plots


ChangeLog
0.10
 * Fixed a problem with vignettes

0.3 January Weiner <january.weiner@gmail.com>
 * pca3d and pca2d return now a data frame that is useful for generating a legend.
 * corrected an error which prevented displaying 1-item groups correctly.


R/pca2d.R
# functions starting with the dot are, by my convention, considered internal
# for this file and the one specific function that is exported from it.
# This is not a formalism, merely a reminder for myself.

.text.frames <- function( coords, text, col= "black", bg= "#66666633" ) {

  sw <- strwidth( text ) * 1.5
  sh <- strheight( text ) * 1.7

  rect( 
        coords[,1] - sw / 2, 
        coords[,2] - sh  / 2,
        coords[,1] + sw / 2, 
        coords[,2] + sh  / 2,
        col= bg,
        border= F
        
        )
  text( coords[,1], coords[,2], text, col= col )

}

.biplot.2D <- function( biplot.coords, biplot.vars ) {

  r <- c( -1.1, 1.1 ) * max( abs( range( biplot.coords[,1:2] ) ) )
  par( usr= rep( r, 2 ) )
  biplot.coords <- biplot.coords[ biplot.vars, ]

  nb <- nrow( biplot.coords )

  arrows( rep( 0, nb ), rep( 0, nb ), biplot.coords[,1], biplot.coords[,2], col= "red", lwd= 2, angle= 20, length= 0.1 )
  labels <- rownames( biplot.coords )

  .text.frames( biplot.coords, labels )

}

.group.labels.2D <- function( coords, group, col= "black", ... ) {
   centr.coords <- calc.centroids( coords, group )
   .text.frames( centr.coords, levels( group ), col= col )
}

.centroids.2D <- function( coords, group, group.shape, group.col, radius, col ) {
  centr.coords <- calc.centroids( coords, group )
  points( centr.coords[,1], centr.coords[,2], col= group.col, pch= group.shape, cex = 2 * radius )
  segments( coords[,1], coords[,2], centr.coords[ group, 1], centr.coords[ group, 2 ], col= col )
}




pca2d <-
function( pca, components= 1:2, col= "grey", title= NULL, new= FALSE,
  axes.color = "black",
  bg= "white",
  radius= NULL,
  group= NULL,
  shape= NULL,
  palette= NULL,
  fancy= FALSE,
  biplot=FALSE,
  biplot.vars= 5,
  show.scale=FALSE,
  show.labels= FALSE,
  labels.col= "black",
  show.axes= TRUE,
  show.axe.titles= TRUE,
  axe.titles= NULL,
  show.plane= TRUE,
  show.shadows= FALSE,
  show.centroids= FALSE,
  show.group.labels= FALSE,
  ...
  ) {

  if( fancy ) {
    show.labels       <- TRUE
    show.shadows      <- TRUE
    show.centroids    <- TRUE
    show.group.labels <- TRUE
  }

  range <- list( x= c( -1, 1 ), y= c( -1, 1 ) )

  if( missing( palette ) ) palette <- palette()

  if( length( components ) > 2 ) {
    cat( "Warning: too many components specified, using only first two" )
    components <- components[1:2]
  }
    
  if( length( components ) != 2 ) {
    stop( "Error: 2 components required" ) 
  }

  pca.coords <- get.pca.coords( pca )

  if( ! missing( biplot ) ) {
    biplot.coords <- get.biplot.coords( pca, biplot )
    biplot.coords <- biplot.coords[ , components]
    biplot <- TRUE
    biplot.vars <- get.biplot.vars( biplot.coords, biplot.vars )
  }

  if( ncol( pca.coords ) < 2 ) 
    stop( sprintf( "Not enough components: found %d, need at least 2", ncol( pca.coords ) ) )

  pca.coords <- pca.coords[ , components ]
  n.p <- nrow( pca.coords )

  ret <- NULL

  if( ! missing( group ) ) {
    group <- factor( group )
    g.n   <- as.numeric( group )
    
    # choose color automatically
    if( missing( col ) ) 
      col <- palette[ g.n %% length( palette ) + 1 ]

    # choose shape automatically
    if( missing( shape ) )
      shape <- c( 15:20 )[ g.n %% 6 + 1 ]

    group.col            <- col[ match( levels( group ), group ) ]
    names( group.col )   <- levels( group )
    group.shape          <- shape[ match( levels( group ), group ) ]
    names( group.shape ) <- levels( group )

    ret <- print.legend( group, group.col, group.shape )
  }

  if( length( shape ) > 1 ) {
    if( length( shape ) != n.p ) 
      stop( sprintf( "Incorrect length of the shape vector (should be %d, is %d)", n.p, length( shape ) ) )
  } else {
    shape <- rep( shape, n.p ) 
  }

  #if( ! all( shape %in% all.shapes ) ) 
    #stop( sprintf( "Incorrect shapes (must be one of c(%s))", paste( all.shapes, collapse= ", " ) ) )

  if( length( col ) > 1 ) {
    if( length( col ) != n.p ) 
      stop( sprintf( "Incorrect length of the color vector (should be %d)", n.p ) )
  } else {
    col <- rep( col, n.p )
  }

  if( ! missing( show.centroids ) & missing( group ) ) 
      stop( "show.centroids can only be used if groups parameter is provided" )

  if( missing( radius ) ) radius <- 1


# initialize the rgl window, clear etc.
  pca_plot_init <- function( ) {
  
    # open new window if requested or not opened yet
    if( new ) {
      cat( "Creating new device\n" )
      #open3d( antialias= 3 )
    }

    if( biplot ) {
      range$x <- c( -1, 1 ) * max( abs( range( pca.coords[,1] ) ) )
      range$y <- c( -1, 1 ) * max( abs( range( pca.coords[,2] ) ) )
    } else {
      range$x <- range( pca.coords[,1] )
      range$y <- range( pca.coords[,2] )
    }


    if( is.null( axe.titles ) ) {
      xlab= paste( "PC", components[1] ) 
      ylab= paste( "PC", components[2] )
    } else {
      xlab= axe.titles[1] 
      ylab= axe.titles[2] ;
    }


    plot( NULL, type= "n", 
      xlim= range$x,
      ylim= range$y,
      xlab= xlab,
      ylab= ylab
       ) 

    usr <- par( "usr" )
    rect( usr[1], usr[3], usr[2], usr[4], col= bg )

    if( show.plane ) {
      abline( h= 0, col= "grey" )
      abline( v= 0, col= "grey" )
    }

  }

  pca_plot_init() 

  # draw shadows ("lollipop" plot)
  if( show.shadows ) 
    segments( pca.coords[,1], pca.coords[,2], pca.coords[,1], 0, col= "grey" )

  points( pca.coords[,1], pca.coords[,2], col= col, pch= shape, cex= radius, ... )

  if( show.centroids )    .centroids.2D( pca.coords, group, group.shape, group.col, 2 * radius, col )
  if( show.group.labels ) .group.labels.2D( pca.coords, group, col= group.col )

  if( ! missing( show.labels ) ) {
    if( class( show.labels ) == "logical" & length( show.labels ) == 1 ) {
      show.labels <- rownames( pca.coords )
    }

    text( pca.coords[,1], pca.coords[,2], show.labels, col= labels.col )
  }

  if( biplot )            .biplot.2D( biplot.coords, biplot.vars )

  return( invisible( ret ) )
}


R/helpers.R

printf <- function( ... ) print( sprintf( ... ) ) 
catf <- function( ... ) cat( sprintf( ... ) ) 

calc.centroids <- function( coords, group ) {

  centr.coords <- apply( coords, 2, function( x ) tapply( x, group, mean ) )

  return( centr.coords )
}

get.pca.coords <- function( pca ) {

  if( class( pca ) == "prcomp" )   return( pca$x )
  if( class( pca ) == "princomp" ) return( pca$scores )
  if( class( pca ) != "matrix" ) stop( "pca must be either a matrix or a PCA object (prcomp, princomp etc.)" )
  return( pca )
}

get.biplot.coords <- function( pca, biplot ) {

  if( class( biplot ) == "matrix" ) return( biplot )
  if( class( pca ) == "prcomp" )    return( pca$rotation )
  if( class( pca ) == "princomp" )  return( pca$loadings )

  stop( "For a biplot, another matrix or a prcomp object as pca is needed")

}

get.biplot.vars <- function( biplot.coords, biplot.vars ) {

  if( length( biplot.vars ) > 1 ) 
    return( biplot.vars )

  if( biplot.vars > nrow( biplot.coords ) ) biplot.vars <- nrow( biplot.coords )

  res <- c()
  for( i in 1:ncol( biplot.coords ) ) {
    res <- c( res, head( order( abs( biplot.coords[,i]), decreasing= TRUE ), biplot.vars ) )
  }

  return( unique( res ) )
}

print.legend <- function( group, group.col, group.shape ) {

  cat( "\nLegend:\n" )
  l.g <- max( nchar( c( levels( group ), unique( group.col ), unique( group.shape ) ) ) )
  fmt <- sprintf( "%% %ds", l.g )
  fmt <- sprintf( "%s: %s, %s\n", fmt, fmt, fmt )
  catf( "%s\n", paste( rep( "-", 3*l.g + 4 ), collapse= "" ) )
  catf( fmt, "group", "color", "shape" )
  catf( "%s\n", paste( rep( "-", 3*l.g + 4 ), collapse= "" ) )
  for( i in 1:length( levels( group ) ) ) {
    catf( fmt, levels( group )[i], group.col[i], group.shape[i])
  }
  cat( "\n" )
}


R/objects3d.R
# draw a series of tetrahaedrons
tetrahaedrons3d <- function( coords, radius= c( 1, 1, 1 ), col= "grey", ... ) {
  coords.n <- NULL
  
  for( r in 1:nrow( coords ) ) {
    p <- coords[ r, ] 

    # ABC
    coords.n <- rbind( coords.n, p + c( -radius[1], 0, -radius[3]/sqrt(2) ) ) # A
    coords.n <- rbind( coords.n, p + c(  radius[1], 0, -radius[3]/sqrt(2) ) ) # B
    coords.n <- rbind( coords.n, p + c(  0,  radius[2], radius[3]/sqrt(2) ) )  # C

    # ABD
    coords.n <- rbind( coords.n, p + c( -radius[1], 0, -radius[3]/sqrt(2) ) ) # A
    coords.n <- rbind( coords.n, p + c(  radius[1], 0, -radius[3]/sqrt(2) ) ) # B
    coords.n <- rbind( coords.n, p + c(  0, -radius[2], radius[3]/sqrt(2) ) )  # D

    # ACD
    coords.n <- rbind( coords.n, p + c( -radius[1], 0, -radius[3]/sqrt(2) ) ) # A
    coords.n <- rbind( coords.n, p + c(  0,  radius[2], radius[3]/sqrt(2) ) )  # C
    coords.n <- rbind( coords.n, p + c(  0, -radius[2], radius[3]/sqrt(2) ) )  # D

    # BCD
    coords.n <- rbind( coords.n, p + c(  radius[1], 0, -radius[3]/sqrt(2) ) ) # B
    coords.n <- rbind( coords.n, p + c(  0,  radius[2], radius[3]/sqrt(2) ) )  # C
    coords.n <- rbind( coords.n, p + c(  0, -radius[2], radius[3]/sqrt(2) ) )  # D
  }

  triangles3d( coords.n, col= col, ... )

}



cubes3d <- function( coords, radius= c( 1, 1, 1), col= "grey", ... ) {
  coords.n <- NULL
  r <- 2 * radius / 3 
  for( i in 1:nrow( coords ) ) {
    p <- coords[ i, ]

    for( j in 1:3 ) {
      other <- (1:3)[ -j ]
      oa <- other[1]
      ob <- other[2]
      for( t in c( -r[j], r[j] ) ) {
        for( q in c( -1, 1 ) ) {
          tmp <- matrix( 0, nrow= 3, ncol= 3 )
          tmp[ , oa ] <- c( q * r[oa], q * r[oa], -q * r[oa ] )
          tmp[ , ob ] <- c( q * r[ob], -q * r[ob], q * r[ob ] )
          tmp[ , j  ] <- rep( t, 3 )
          tmp <- t( apply( tmp, 1, function( x ) p + x ) )
          coords.n <- rbind( coords.n, tmp )
        }
      }
    }
  }       
  triangles3d( coords.n, col= col, ... )
}

# return the basic cone mesh
# scale is necessary because of the dependence on the aspect ratio
.getcone <- function( r, h, scale= NULL ) {

  n  <- length( .sin.t )
  xv <- r * .sin.t
  yv <- rep( 0, n )
  zv <- r * .cos.t

  if( missing( scale ) ) scale <- rep( 1, 3 )

  scale <- 1 / scale
  sx <- scale[1]
  sy <- scale[2]
  sz <- scale[3]

  tmp <- NULL
  for( i in 1:(n-1) ) {
    tmp <- rbind( tmp,
      c( 0, 0, 0 ),
      scale3d( c( xv[i],   yv[i],   zv[i]   ), sx, sy, sz ),
      scale3d( c( xv[i+1], yv[i+1], zv[i+1] ), sx, sy, sz ) )
  }
  for( i in 1:(n-1) ) {
    tmp <- rbind( tmp,
      c( 0, h, 0 ),
      scale3d( c( xv[i],   yv[i],   zv[i]   ), sx, sy, sz ),
      scale3d( c( xv[i+1], yv[i+1], zv[i+1] ), sx, sy, sz ) )
  }
  tmp
}

# vector cross product
.cross3 <- function(a,b) {
  c(a[2]*b[3]-a[3]*b[2], -a[1]*b[3]+a[3]*b[1], a[1]*b[2]-a[2]*b[1])
}

# draw a cone (e.g. tip of an arrow)
cone3d <- function( base, tip, radius= 10, col= "grey", scale= NULL, ... ) {
  start <- rep( 0, 3 )

  if( missing( scale ) ) scale= rep( 1, 0 ) 
  else scale <- max( scale ) / scale


  tip  <- as.vector( tip ) * scale
  base <- as.vector( base ) * scale

  v1 <- tip
  v2 <- c( 0, 100, 0 )
  o <- .cross3( v1, v2 )
  theta <- acos( sum( v1 * v2 ) / ( sqrt(sum( v1  *  v1 )) * sqrt(sum( v2  *  v2 )) ) )
  vl <- sqrt( sum( tip^2 ) )

  tmp <- .getcone( radius, vl )
  tmp <- translate3d( rotate3d( tmp, theta, o[1], o[2], o[3] ), base[1], base[2], base[3] )
  scale <- 1 / scale
  tmp <- t( apply( tmp, 1, function( x ) x * scale ) )
  triangles3d( tmp, col= col, ... )
}




arrows3d <- function( coords, headlength= 0.035, head= "end", scale= NULL, radius = NULL, ... ) {

  head <- match.arg( head, c( "start", "end", "both" ) )
  narr <- nrow( coords ) / 2 
  n    <- nrow( coords )

  starts <- coords[ seq( 1, n, by= 2 ), ]
  ends   <- coords[ seq( 2, n, by= 2 ), ]
  if( missing( radius ) ) radius <- ( max( coords ) - min( coords ) ) / 50

  segments3d( coords, ... )
  if( head == "end" | head == "both" ) {
    for( i in 1:narr ) {
      s <- starts[i,]
      e <- ends[i,]
      base <- e - ( e - s ) * headlength 
      tip  <- ( e - s ) * headlength
      cone3d( base, tip, radius= radius, scale= scale, ... )

    }
  }

}


R/pca3d.R
  obsolete_cone3d <- function( base=c(0,0,0),
                      tip=c(0,0,1),
                      rad=1,
                      n=30,
                      draw.base=TRUE,qmesh=FALSE,
                      trans = par3d("userMatrix"), ...) {
   ax <- tip-base
   if (missing(trans) && !rgl.cur()) trans <- diag(4)
   ### is there a better way?
   if (ax[1]!=0) {
     p1 <- c(-ax[2]/ax[1],1,0)
     p1 <- p1/sqrt(sum(p1^2))
     if (p1[1]!=0) {
       p2 <- c(-p1[2]/p1[1],1,0)
       p2[3] <- -sum(p2*ax)
       p2 <- p2/sqrt(sum(p2^2))
     } else {
       p2 <- c(0,0,1)
     }
   } else if (ax[2]!=0) {
     p1 <- c(0,-ax[3]/ax[2],1)
     p1 <- p1/sqrt(sum(p1^2))
     if (p1[1]!=0) {
       p2 <- c(0,-p1[3]/p1[2],1)
       p2[3] <- -sum(p2*ax)
       p2 <- p2/sqrt(sum(p2^2))
     } else {
       p2 <- c(1,0,0)
     }
   } else {
     p1 <- c(0,1,0); p2 <- c(1,0,0)
   }
   degvec <- seq(0,2*pi,length=n+1)[-1]
   ecoord2 <- function(theta) {
     base+rad*(cos(theta)*p1+sin(theta)*p2)
   }
   i <- rbind(1:n,c(2:n,1),rep(n+1,n))
   v <- cbind(sapply(degvec,ecoord2),tip)
   if (qmesh) 
     ## minor kluge for quads -- draw tip twice
     i <- rbind(i,rep(n+1,n))
   if (draw.base) {
     v <- cbind(v,base)
     i.x <- rbind(c(2:n,1),1:n,rep(n+2,n))
     if (qmesh)  ## add base twice
       i.x <-  rbind(i.x,rep(n+2,n))
     i <- cbind(i,i.x)
   }
   if (qmesh) v <- rbind(v,rep(1,ncol(v))) ## homogeneous
   if (!qmesh)
     triangles3d(v[1,i],v[2,i],v[3,i],...)
   else
     return(rotate3d(qmesh3d(v,i,material=...), matrix=trans))
 }      



# functions starting with the dot are, by my convention, considered internal
# for this file and the one specific function that is exported from it.
# This is not a formalism, merely a reminder for myself.

# for drawing circles in 3D, precalculate some values
.sin.t <- sin( seq( 0, 2 * pi, len= 10 ) )
.cos.t <- cos( seq( 0, 2 * pi, len= 10 ) )


# draw labels for each centroid
.group.labels <- function( coords, group, ... ) {
   centr.coords <- calc.centroids( coords, group )
   texts3d( centr.coords[,1], centr.coords[,2], centr.coords[,3], levels( group ), col= "black", cex= 2, ... )
}

.show.axe.titles <- function( r, show.scale, axes.color, components ) {
      titles = paste( "PC", components )
      
      texts3d( 
        c( 1.1*r[ 2,1 ], 0, 0 ),
        c( 0, 1.1*r[ 2,2 ], 0 ),
        c( 0, 0, 1.1*r[ 2,3 ] ), col= "black", titles ) 

    
    if( show.scale ) {
      axis3d('x',pos=c( NA, 0, 0 ), col= axes.color)
      axis3d('y',pos=c( 0, NA, 0 ), col= axes.color)
      axis3d('z',pos=c( 0, 0, NA ), col= axes.color)
    }
}

.show.plane <- function( r ) {
      quads3d( 
        rbind( 
          c( r[1,1],0,r[1,3] ),
          c( r[1,1],0,r[2,3] ),
          c( r[2,1],0,r[2,3] ),
          c( r[2,1],0,r[1,3] ) 
          ),
        col= "grey",
        alpha= 0.2
      ) 

}

.show.axes <- function( axes.color, ranges ) {

      axes <- rbind( 
        c( ranges[1,1], 0, 0 ),
        c( ranges[2,1], 0, 0 ),
        c( 0, ranges[1,2], 0 ),
        c( 0, ranges[2,2], 0 ),
        c( 0, 0, ranges[1,3] ),
        c( 0, 0, ranges[2,3] )
        )
      segments3d( axes, col= axes.color )

      radius <- 10

      scale <- c( 1, 1, 1 )
      if( ! missing( ranges ) ) {
        scale <- ranges[2,]
        radius <- max( scale ) / 50
      }

      arrows3d( axes, radius= radius, scale= scale,  col= axes.color )

     #cone3d( base= c( r[2,1] * 0.95, 0, 0 ), tip= c( r[2,1] * 0.05, 0, 0 ), radius= radius, ranges= r[2,] )
     #cone3d( base= c( 0, r[2,2] * 0.95, 0 ), tip= c( 0, r[2,2] * 0.05, 0 ), radius= radius, scale= r[2,] )
     #cone3d( base= c( 0, 0, r[2,3] * 0.95 ), tip= c( 0, 0, r[2,3] * 0.05 ), radius= radius, scale= r[2,] )

# draw arrows
#     triangles3d( rbind( 
#       c( r[2,1], 0, 0 ),
#       c( r[2,1] * 0.95, r[1,2] * 0.02, 0  ),
#       c( r[2,1] * 0.95, r[2,2] * 0.02, 0  ),
#       c( r[2,1], 0, 0 ), 
#       c( r[2,1] * 0.95, 0, r[1,3] * 0.02 ),
#       c( r[2,1] * 0.95, 0, r[2,3] * 0.02 ),
#       c( 0, r[2,2], 0 ),
#       c( r[1,1] * 0.02, r[2,2] * 0.95, 0  ),
#       c( r[2,1] * 0.02, r[2,2] * 0.95, 0  ),
#
#       c( 0, r[2,2], 0 ),
#       c(0, r[2,2] * 0.95,  r[1,1] * 0.02  ),
#       c(0, r[2,2] * 0.95,  r[2,1] * 0.02  ),
#
#       c( 0, 0, r[2,3] ),
#       c( r[1,1] * 0.02, 0, r[2,3] * 0.95  ),
#       c( r[2,1] * 0.02, 0, r[2,3] * 0.95  ),
#       c( 0, 0, r[2,3] ),
#       c( 0, r[1,2] * 0.02, r[2,3] * 0.95 ),
#       c( 0, r[2,2] * 0.02, r[2,3] * 0.95 ) 
#     ), col= axes.color )


}

.centroids.3D <- function( coords, group, radius, group.shape, group.col, shape.functions ) {

  centr.coords <- calc.centroids( coords, group )

  rownames( centr.coords ) <- levels( group )

  for( l in levels( group ) ) {
    shape.functions[[ group.shape[l] ]]( centr.coords[l,,drop= F], col= group.col[l], radius= radius, alpha= 0.3 )
    #shape.functions[[ centr.shapes[l] ]]( centr.coords[l,], col= centr.cols[ l ], alpha= 0.3, radius= radius, shininess= 100 ) 
    n.l <- length( which( group == l ) )
    tmp <- coords[ group == l, ]

    tmp <- rbind( tmp, t( sapply( 1:nrow( tmp ), function( x ) centr.coords[ l, ] ) ) )
    tmp <- tmp[ as.vector( rbind( 1:n.l, (n.l + 1):(2*n.l) ) ), ]
    segments3d( tmp, col= group.col[ l ] )
  }

}

.draw.shadows <- function( pca.coords, xz.functions, shape, radius) {

  seg <- NULL
  for( i in 1:nrow( pca.coords ) ) {
    tmp <- pca.coords[ i, ]
    seg <- rbind( seg, tmp )
    tmp[ 2 ] <- 0
    seg <- rbind( seg, tmp )
    #squarexz( tmp[1], tmp[2] + 0.001, tmp[3], radius= radius, col= "grey" )
    xz.functions[[ shape[i] ]]( tmp[1], 0 + 0.001 * radius[2], tmp[3], radius= radius, col= "grey" )
    xz.functions[[ shape[i] ]]( tmp[1], 0 - 0.001 * radius[2], tmp[3], radius= radius, col= "grey" )
  }
  segments3d( seg, col= "grey", alpha= 0.5 )
}

.squarexz <- function( x, y, z, radius= c( 1, 1, 1 ), col= "grey", alpha= 0.5 ) {
  tmp <- NULL
  r <- 2*radius/3
  for( i in 1:length( x ) ) {
    tmp <- rbind( tmp,
      c( x[i] - r[1], y[i], z[i] - r[3] ), 
      c( x[i] - r[1], y[i], z[i] + r[3] ), 
      c( x[i] + r[1], y[i], z[i] - r[3] ), 
      c( x[i] + r[1], y[i], z[i] + r[3] ), 
      c( x[i] - r[1], y[i], z[i] + r[3] ), 
      c( x[i] + r[1], y[i], z[i] - r[3] ) )
  }
  triangles3d( tmp, col= col, alpha= alpha )
}

.trianglexz <- function( x, y, z, radius= c( 1, 1, 1 ), col= "grey", alpha= 0.5 ) {
  tmp <- NULL
  for( i in 1:length( x ) ) {
    tmp <- rbind( tmp, 
      c( x[i]                            , y[i], z[i] + radius[3]     ),
      c( x[i] + sqrt( 3 ) * radius[1] / 2, y[i], z[i] - radius[3] / 2 ),
      c( x[i] - sqrt( 3 ) * radius[1] / 2, y[i], z[i] - radius[3] / 2 )
      )
  }
  triangles3d( tmp, col= col, alpha= alpha )

}

.circlexz <- function( x, y, z, radius= c( 1, 1, 1), col= "grey", alpha= 0.5 ) {
  
  r <- radius * 2 / 3 
  n <- length( .sin.t )
  xv <- x + r[1] * .sin.t
  yv <- rep( y, n )
  zv <- z + r[3] * .cos.t

  tmp <- NULL
  for( i in 1:(n-1) ) {
    tmp <- rbind( tmp, 
      c( x, y, z ),
      c( xv[i],   yv[i],   zv[i]   ),
      c( xv[i+1], yv[i+1], zv[i+1] ) )
  }

  triangles3d( tmp, col= col, alpha= alpha )
}

.myspheres3d <- function( coords, radius= c( 1, 1, 1 ), ... ) {
    spheres3d( coords, radius= mean( radius ), ... )
}

.biplot.3D <- function( biplot.coords, biplot.vars, ranges= NULL ) {

  nvar <- length( biplot.vars )

  coords <- matrix( 0, ncol= 3, nrow= nvar * 2 )
  tips <- biplot.coords[ biplot.vars, ]
  coords[ seq( 2, nvar * 2, by= 2 ), ] <- tips

  mx <- max( abs( coords ) )

  coords <- coords / mx 
  scale <- c( 1, 1, 1 )
  if( ! missing( ranges ) ) {
    scale <- ranges[2,]
    coords <- t( apply( coords, 1, function( x ) x * scale ) )
  }

  arrows3d( coords, col= "red", scale= scale )

  if( length( rownames( biplot.coords ) ) > 0 ) 
    names <- rownames( biplot.coords )[biplot.vars]
  else
    names <- biplot.vars

  coords <- coords[ seq( 2, nvar * 2, by = 2 ), ]
  texts3d( coords[,1], coords[,2], coords[,3], names, col= "black" )





}


pca3d <- function( pca, components= 1:3, col= "grey", title= NULL, new= FALSE,
  axes.color = "grey",
  bg= "white",
  radius= NULL,
  group= NULL,
  shape= "sphere",
  palette= NULL,
  fancy= FALSE,
  biplot= FALSE,
  biplot.vars= 5,
  show.scale= FALSE,
  show.labels= FALSE,
  labels.col= "black",
  show.axes= TRUE,
  show.axe.titles= TRUE,
  show.plane= TRUE,
  show.shadows= FALSE,
  show.centroids= FALSE,
  show.group.labels= FALSE
  ) {

  if( fancy ) {
    show.labels       <- TRUE
    show.shadows      <- TRUE
    show.centroids    <- TRUE
    show.group.labels <- TRUE
  }

  all.shapes<- c( "sphere", "tetrahaedron", "cube" )
  if( missing( palette ) ) palette <- palette()

  if( length( components ) != 3 ) 
    stop( "Error: 3 components required" ) 

  pca.coords <- get.pca.coords( pca )

  if( ncol( pca.coords ) < 3 ) 
    stop( sprintf( "Not enough components: found %d, need at least 3", ncol( pca.coords ) ) )

  if( ! missing( biplot ) ) {
    biplot.coords <- get.biplot.coords( pca, biplot )
    biplot.coords <- biplot.coords[ , components]
    biplot <- TRUE
    biplot.vars <- get.biplot.vars( biplot.coords, biplot.vars )
  }


  pca.coords <- pca.coords[ , components ]
  n.p <- nrow( pca.coords )
  main.ranges <- apply( pca.coords, 2, function( x ) c( -max( abs( range( x ) ) ), max( abs( range( x ) ) ) ) )

  if( ! missing( group ) ) {
    group <- factor( group )
    g.n   <- as.numeric( group )
    
    # choose color automatically
    if( missing( col ) ) 
      col <- palette[ g.n %% length( palette ) + 1 ]

    # choose shape automatically
    if( missing( shape ) ) 
      shape <- all.shapes[ g.n %% 3 + 1 ]

    group.col <- col[ match( levels( group ), group ) ]
    names( group.col ) <- levels( group )
    group.shape <- shape[ match( levels( group ), group ) ]
    names( group.shape ) <- levels( group )

    print.legend( group, group.col, group.shape )
    
  }

  if( length( shape ) > 1 ) {
    if( length( shape ) != n.p ) 
      stop( sprintf( "Incorrect length of the shape vector (should be %d, is %d)", n.p, length( shape ) ) )
  } else {
    shape <- rep( shape, n.p ) 
  }

  if( ! all( shape %in% all.shapes ) ) 
    stop( sprintf( "Incorrect shapes (must be one of c(%s))", paste( all.shapes, collapse= ", " ) ) )

  if( length( col ) > 1 ) {
    if( length( col ) != n.p ) 
      stop( sprintf( "Incorrect length of the color vector (should be %d)", n.p ) )
  } else {
    col <- rep( col, n.p )
  }

  if( ! missing( show.centroids ) & missing( group ) ) 
      stop( "show.centroids can only be used if groups parameter is provided" )

  if( missing( radius ) ) radius <- 1
  radius <- rep( radius, 3 )

  for( i in 1:3 ) {
    r <- 2*max( abs( pca.coords[ , i ] ) )
    radius[i] <- radius[i] * r / 100
  }


  # the following three functions are for throwing shadows of objects on the
  # XZ plane
# initialize the rgl window, clear etc.
  pca_plot_init <- function( ) {
  
    # open new window if requested or not opened yet
    if( new | rgl.cur() == 0 ) {
      cat( "Creating new device\n" )
      rgl.open()
      #open3d( antialias= 3 )
      par3d( windowRect= 50 + c( 0, 0, 640, 640 ) )
      rgl.viewpoint( theta= 45, phi= 30, fov= 60, zoom= 1 )
      bg3d( bg )
      rgl.pop( "lights" )
      light3d( specular= "black" )
    }

    par3d( skipRedraw= TRUE ) 

    rgl.clear()

    #r <- apply( pca.coords, 2, function( x ) c( -max( abs( range( x ) ) ), max( abs( range( x ) ) ) ) )
    if( show.plane ) .show.plane( main.ranges )

    if( show.axes ) .show.axes( axes.color, main.ranges * 1.1 )

    if( show.axe.titles ) .show.axe.titles( main.ranges, show.scale, axes.color, components )
    
    if( show.scale ) {

      axis3d('x',pos=c( NA, 0, 0 ))
      axis3d('y',pos=c( 0, NA, 0 ))
      axis3d('z',pos=c( 0, 0, NA ))


    }
  }


# finalize the plot
  pca_plot_finish <- function( ) {
    aspect3d( c( 1, 1, 1 ) )
    #aspect3d( "iso" )
    par3d( skipRedraw= F )
    rgl.bringtotop()
  }


  shape.functions <- list( sphere= .myspheres3d, tetrahaedron= tetrahaedrons3d, cube= cubes3d )
  xz.functions <- list( sphere= .circlexz, tetrahaedron= .trianglexz, cube= .squarexz )

  plot.scale <- pca_plot_init() 

  # draw shadows ("lollipop" plot)
  if( show.shadows )   .draw.shadows( pca.coords, xz.functions, shape, radius )
  if( show.centroids ) .centroids.3D( pca.coords, group, 2 * radius, group.shape, group.col, shape.functions )
  if( show.group.labels ) .group.labels( pca.coords, group )

  for( s in unique( shape ) ) {
    for( c in unique( col ) ) {
      sel <- which( shape == s & col == c )
      if( length( sel ) > 0 ) {
        shape.functions[[s]]( pca.coords[ sel, ], col= as.character( c ), alpha= 1, radius= radius, shininess= 100 ) 
      }
    }
  }

  if( biplot ) 
    .biplot.3D( biplot.coords, biplot.vars, main.ranges )

  if( ! missing( show.labels ) ) {
    if( class( show.labels ) == "logical" & length( show.labels ) == 1 ) {
      show.labels <- rownames( pca.coords )
    }
    texts3d( pca.coords[,1], pca.coords[,2], pca.coords[,3], show.labels, col= labels.col )
  }

  pca_plot_finish() 
}
